File: backend/app/schemas/sources.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict


class SourceSystemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    name: str
    description: str | None
    source_type: str
    is_active: bool
    created_at: datetime
    updated_at: datetime


class SourceSystemPage(BaseModel):
    items: list[SourceSystemResponse]
    total: int
    page: int
    page_size: int


class SourceFileResponse(BaseModel):
    id: int
    source_system_id: int
    source_system_code: str
    original_filename: str
    stored_filename: str
    relative_path: str
    file_extension: str
    mime_type: str
    file_size_bytes: int
    sha256_checksum: str
    status: str
    discovered_at: datetime
    registered_at: datetime
    created_at: datetime
    updated_at: datetime


class SourceFilePage(BaseModel):
    items: list[SourceFileResponse]
    total: int
    page: int
    page_size: int


class PipelineRunStepResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    step_name: str
    step_order: int
    status: str
    started_at: datetime
    completed_at: datetime | None
    metadata_json: dict[str, Any] | None
    error_message: str | None


class PipelineRunResponse(BaseModel):
    id: int
    run_type: str
    status: str
    started_at: datetime
    completed_at: datetime | None
    source_file_id: int | None
    records_extracted: int
    records_accepted: int
    records_rejected: int
    error_message: str | None
    created_at: datetime
    updated_at: datetime
    steps: list[PipelineRunStepResponse]


class PipelineRunPage(BaseModel):
    items: list[PipelineRunResponse]
    total: int
    page: int
    page_size: int


class UploadSuccessResponse(BaseModel):
    status: Literal["registered"]
    source_file_id: int
    original_filename: str
    stored_filename: str
    sha256_checksum: str
    file_size_bytes: int
    pipeline_run_id: int


class UploadDuplicateResponse(BaseModel):
    status: Literal["duplicate"]
    message: str
    existing_source_file_id: int
    sha256_checksum: str
    pipeline_run_id: int


class UploadErrorResponse(BaseModel):
    status: Literal["validation_error", "failed"]
    code: str
    message: str
    pipeline_run_id: int | None

File: frontend/src/api/sources.ts
import { API_BASE_URL } from "./health";

export interface SourceSystem {
  id: number;
  code: string;
  name: string;
  description: string | null;
  source_type: string;
  is_active: boolean;
}

export interface SourceFile {
  id: number;
  source_system_id: number;
  source_system_code: string;
  original_filename: string;
  stored_filename: string;
  relative_path: string;
  file_extension: string;
  mime_type: string;
  file_size_bytes: number;
  sha256_checksum: string;
  status: string;
  registered_at: string;
}

interface Page<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
}

export interface UploadResult {
  status: "registered" | "duplicate" | "validation_error" | "failed";
  message?: string;
  source_file_id?: number;
  existing_source_file_id?: number;
  original_filename?: string;
  stored_filename?: string;
  sha256_checksum?: string;
  file_size_bytes?: number;
  pipeline_run_id?: number;
  code?: string;
}

async function parseResponse<T>(response: Response): Promise<T> {
  const body = (await response.json()) as T;
  if (!response.ok) {
    const failure = body as UploadResult;
    throw new Error(failure.message ?? `Request failed with status ${response.status}`);
  }
  return body;
}

export async function fetchSourceSystems(signal?: AbortSignal): Promise<SourceSystem[]> {
  const response = await fetch(`${API_BASE_URL}/source-systems?page_size=100`, { signal });
  return (await parseResponse<Page<SourceSystem>>(response)).items;
}

export async function fetchSourceFiles(signal?: AbortSignal): Promise<SourceFile[]> {
  const response = await fetch(`${API_BASE_URL}/source-files?page_size=20`, { signal });
  return (await parseResponse<Page<SourceFile>>(response)).items;
}

export async function fetchSourceFile(sourceFileId: number): Promise<SourceFile> {
  const response = await fetch(`${API_BASE_URL}/source-files/${sourceFileId}`);
  return parseResponse<SourceFile>(response);
}

export async function uploadSourceFile(
  file: File,
  sourceSystemCode: string,
): Promise<UploadResult> {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("source_system_code", sourceSystemCode);
  const response = await fetch(`${API_BASE_URL}/source-files/upload`, {
    method: "POST",
    body: formData,
  });
  return parseResponse<UploadResult>(response);
}
